docs: add 'Ask users for dates and times' pattern page (#2080)
agriculturegovau/agds-next: docs/content/patterns/ask-users-for-dates-and-times/index.mdx

---
title: Ask users for dates and times
description: When we ask for date, time or a combination of both there are a range of components that can be used to meet user needs in different situations.
relatedComponents: ['time-input', 'time-picker']
---

## Time input

[Time input](/components/time-input) allows users to type an exact time - for example, the known time of an incident or event.

When the user enters a time, formatting is applied to collect input in a consistent format.

> It should be used in situations when user know and can provide an exact and arbitrary time.

Reviewer: ```suggestion
It should be used in situations when a user knows and can provide an exact and arbitrary time.
```

		/>
	);
};
```

## Time picker

[Time picker](/components/time-picker) is a composition of [combobox](/components/combobox) that allows users to select a time from a predefined list of consistent time intervals.

As the user enters text into the field, autocomplete narrows the list to match their input.

The list is closed when a time is selected.

It should be used when a user needs to select from a large list of specific times.

```jsx live
() => {
	const [value, setValue] = React.useState(null);

	const onChange = (timeValue) => {
		console.log('onChange', timeValue);
		setValue(timeValue);
	};

	return (
		<TimePicker
			hideOptionalLabel
			hint="Start typing to see results"
			label="Select a time"
			onChange={onChange}
			value={value}
		/>
	);
};
```

## Asking for specific known dates

When a specific known date is required use a format tolerant [Text input](/components/text-input) that validates acceptable formats and auto formats minor issues like spaces or capital letters.

Examples of a specific known date are birth date, when an application was made or an expiry date.

```jsx live
<TextInput
	hideOptionalLabel
	hint="(e.g. 15/02/1982)"
	label="Date of event"

> 	type="date"

Reviewer: We shouldn't be using `type="date"` right? We don't want the browser's native calendar

/>
```

## Asking for approximate dates

When asking for a date that may be general or hard for users to remember, allow them to provide an approximate date.

Use a format-tolerant text input that accepts a range of valid formats and automatically reformats minor issues, such as extra spaces or capital letters.

Examples of approximate dates include any dates that are indicative and non-specific like ’January 2025’.

```jsx live
<TextInput
	hideOptionalLabel
	hint="(e.g. December 2025)"
	label="Approximate date"
/>
```

## Picking dates

> Use [Date picker](/components/date-picker-next) when a user needs to pick a specific or available date from a calendar.

Reviewer: This isn't true. They don't need to pick a date from a calendar and are not forced to. DatePickerNext should be an asking for specific known dates imo as an alternative

Reviewer 2: Use [Date picker](/components/date-picker-next) so a user can pick a specific or available date from a calendar.
?

Author: Updated in [9bd6fd](https://github.com/agriculturegovau/agds-next/pull/2080/commits/9bd6fd0b4139c1144886773d54702512c9b4b9c8)

The Date picker calendar allows selection of both past and future dates. It displays one month at a time, with controls to navigate by month or year.

Use Date picker when a user needs to pick an unknown date or when they need to find an available date like when booking appointments.

```jsx live
() => {
	const [value, setValue] = React.useState();
	return (
		<DatePickerNext
			hideOptionalLabel
			label="Select date"
			onChange={setValue}
			value={value}
		/>
	);
};
```

> ## Picking date spans

Reviewer: Why wouldn't we be consistent with the name of the component? Range is also much clearer than Span
```suggestion
## Picking date ranges
```

The Date range picker calendars allow selection of both past and future dates. They display two months at a time on large screens, with controls to navigate by month or year.

Use Date range picker when a user needs to pick a start and end. For example, when providing transport times or booking accommodation.

```jsx live
() => {
	const [value, setValue] = React.useState({ from: undefined, to: undefined });
	return (
		<DateRangePickerNext value={value} onChange={setValue} hideOptionalLabel />
	);
};
```

## Picking a date and time

Use the [Grouped fields](/components/grouped-fields) component to combine closely related inputs like [Date picker](/components/date-picker-next) and [Time picker](/components/time-picker) when asking users to provide a specific date and time.

An examples of when we combine date and time in Grouped fields is booking a departure time or an appointment.

```jsx live
() => {
	const [dateValue, setDateValue] = React.useState();
	const [timeValue, setTimeValue] = React.useState();
	return (
		<GroupedFields legend="Departure time" hideOptionalLabel>
			{({ field1Props, field2Props }) => (
				<>
					<DatePickerNext
						hideOptionalLabel

> 						label="Select date"

Reviewer: ```suggestion
						label="Date"
```

						onChange={setDateValue}
						value={dateValue}
						{...field1Props}
					/>
					<TimePicker
						hideOptionalLabel

> 						label="Select available time"

Reviewer: ```suggestion
						label="Select time"
```

						onChange={setTimeValue}
						value={timeValue}
						{...field2Props}
					/>
				</>
			)}
		</GroupedFields>
	);
};
```

## Time zones

Use the [Grouped fields](/components/grouped-fields) component to combine closely related inputs like [Time picker](/components/time-picker) and a time zone [Select](/components/select) when asking users for a time and the time zone is important.

An example of when we combine time and time zone in Grouped fields is when the time a consignment will arrive in another country is required.

```jsx live
() => {
	const [timeValue, setTimeValue] = React.useState();
	const [timezoneValue, setTimezoneValue] = React.useState();
	return (
		<GroupedFields legend="Departure time" hideOptionalLabel>
			{({ field1Props, field2Props }) => (
				<>
					<TimePicker
						hideOptionalLabel
						label="Time"
						message="Enter a valid time"
						onChange={setTimeValue}
						value={timeValue}
						{...field1Props}
					/>
					<Select
						hideOptionalLabel

> 						label="Select time"

Reviewer: ```suggestion
						label="Timezone"
```
